### ISSUE TITLE
Cannot access error payloads from REST API responses

### ISSUE THREAD

**Issue by chrisbonifacio** [aws-amplify/amplify-js#12730]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-js/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-js/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### JavaScript Framework

Not applicable

### Amplify APIs

REST API

### Amplify Categories

api

### Environment information

<details>

```
# Put output below this line


```

</details>

### Describe the bug

When receiving and handling error status codes from a REST API, the thrown error does not expose the response payload. Instead, it simply throws `Unknown error`

### Expected behavior

The error thrown should include the response payload from the server

### Reproduction steps

1. `git clone git@github.com:chrisbonifacio/amplify-v6-pages-router.git && cd amplify-v6-pages-router`
2. `npm i`
3. `amplify init`
4. `amplify push -y`
5. `npm run dev`
6. inspect page > Console
7. click `Get Items` to send request to REST API and receive a response with a 500 status code
8. observe `Unknown error` thrown within the console

### Code Snippet

```javascript
const getItems = async () => {
    try {
      const req = post({
        apiName: "api669dbb52",
        path: "/items",
        options: {
          headers: {
            Authorization: "test",
          },
          body: {
            statusCode: 500,
          },
        },
      });

      const res = await req.response;

      console.log(await res.body.json());
    } catch (error) {
      console.log(error);
    }
  };
```

### Log output

<details>

```
index.tsx:31 

POST https://6aoffdgx2e.execute-api.us-east-1.amazonaws.com/dev/items 429 (Too Many Requests)
fetchTransferHandler @ fetch.mjs:15
composedHandler @ composeTransferHandler.mjs:17
retryMiddleware @ middleware.mjs:32
await in retryMiddleware (async)
userAgentMiddleware @ middleware.mjs:16
eval @ composeTransferHandler.mjs:23
transferHandler @ handler.mjs:75
eval @ publicApis.mjs:44
await in eval (async)
job @ createCancellableOperation.mjs:33
createCancellableOperation @ createCancellableOperation.mjs:69
publicHandler @ publicApis.mjs:27
post @ publicApis.mjs:53
post @ index.mjs:96
getItems @ index.tsx:31
callCallback @ react-dom.development.js:4164
invokeGuardedCallbackDev @ react-dom.development.js:4213
invokeGuardedCallback @ react-dom.development.js:4277
invokeGuardedCallbackAndCatchFirstError @ react-dom.development.js:4291
executeDispatch @ react-dom.development.js:9041
processDispatchQueueItemsInOrder @ react-dom.development.js:9073
processDispatchQueue @ react-dom.development.js:9086
dispatchEventsForPlugins @ react-dom.development.js:9097
eval @ react-dom.development.js:9288
batchedUpdates$1 @ react-dom.development.js:26135
batchedUpdates @ react-dom.development.js:3991
dispatchEventForPluginEventSystem @ react-dom.development.js:9287
dispatchEventWithEnableCapturePhaseSelectiveHydrationWithoutDiscreteEventReplay @ react-dom.development.js:6465
dispatchEvent @ react-dom.development.js:6457
dispatchDiscreteEvent @ react-dom.development.js:6430
Show 27 more frames
Show less
index.tsx:50 UnknownError: Unknown error
    at buildRestApiServiceError (serviceError.mjs:19:26)
    at parseRestApiServiceError (serviceError.mjs:32:26)
    at async job (createCancellableOperation.mjs:35:23)
    at async getItems (index.tsx:44:19)
```

</details>

### aws-exports.js

{
  "aws_project_region": "us-east-1",
  "aws_appsync_graphqlEndpoint": "https://svhwm4zmqjem3ohpd7t3hfvxli.appsync-api.us-east-1.amazonaws.com/graphql",
  "aws_appsync_region": "us-east-1",
  "aws_appsync_authenticationType": "API_KEY",
  "aws_appsync_apiKey": "da2-datczwoqqnbjlg3anti7tpjcma",
  "aws_cloud_logic_custom": [
    {
      "name": "api669dbb52",
      "endpoint": "https://6aoffdgx2e.execute-api.us-east-1.amazonaws.com/dev",
      "region": "us-east-1"
    }
  ]
}

**Comment by Dung288-806** [comment]
Hmmm. I'm facing this issue.

**Comment by arianshirepublik** [comment]
I'm facing this issue.  Can the issue be solved as soon as possible?

**Comment by 40818419** [comment]
guys, facing an issue with strings as error messages not being parsed... 

- i would really like to avoid parsing this error into something readable, even then error code would not be accessible...
- is there a plan to support strings?
 
<img width="566" alt="Screenshot 2023-12-20 at 12 40 29" src="https://github.com/aws-amplify/amplify-js/assets/2235499/c41396ad-8613-4869-9b73-50edfc7da9e4">

**Comment by cwomack** [comment]
@Dung288-806 and @40818419, we are currently investigating this and will provide an update as soon as we can.

**Comment by MiqueiasGFernandes** [comment]
> @Dung288-806 and @40818419, we are currently investigating this and will provide an update as soon as we can.

@cwomack Could I contribute for somenthing in this investigation?

**Comment by brettbs** [comment]
Also running into this

**Comment by 40818419** [comment]
> @Dung288-806 and @40818419, we are currently investigating this and will provide an update as soon as we can.

let me know, if you need any assistance in testing RC

**Comment by cwomack** [comment]
@40818419, appreciate it and will let you know! 

@MiqueiasGFernandes, we are working on a fix for this and will update this issue as soon as we can. Thank you for offering to help contribute!

**Comment by kingferiol** [comment]
Same issue here! Thanks for any help.

**Comment by Dung288-806** [comment]
Is it done @cwomack ?

**Comment by cwomack** [comment]
@Dung288-806, the fix is merged and just pending release at this point.  We plan on doing one today, but will be sure to comment and confirm once it's out!

**Comment by jimblanc** [comment]
The fix for this issue has been released with [Amplify JS v6.0.15](https://www.npmjs.com/package/aws-amplify/v/6.0.15). Please let us know if you're run into any other issues! Associated documentation can be found here: https://docs.amplify.aws/react/build-a-backend/restapi/fetch-data/#access-http-response-from-errors
